GIC driver refactor

sandeepbrcm opened the pull request:
Refactor GIC drivers to use common code for GICv1/v2/v3  especially for distributor config.
This reduces the code base to some extent and does some cleanups.

Commit:
drivers: interrupt_controller: refactor GIC driver

GICv3 and GICv2 GICv1 drivers are refactored to use common functions
for distributor interface. Headers are reorgalized accordingly.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

Commit:
CODEOWNERS: update gic file list

Update gic files list.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

sandeepbrcm commented:
some failures are there with GICv1.  So I will keep fixing it here as local builds I might miss to cover all.

sandeepbrcm commented:
@stephanosio 
tests/kernel/interrupt/arch.interrupt in qemu_arc_em:tests/kernel/interrupt/arch.interrupt
tests/kernel/interrupt/arch.interrupt in qemu_arc_hs:tests/kernel/interrupt/arch.interrupt

These two failures are unrelated to GIC.  qemu_arc_hs/em uses some other interrupt controller.
I am not sure why they are failing.  Any idea ?

sandeepbrcm commented on include/arch/arm/aarch64/cpu.h:
```diff
@@ -171,6 +171,7 @@
 		(((mpidr) >> MPIDR_AFF##aff_level##_SHIFT) & MPIDR_AFFLVL_MASK)
 
-#define GET_MPIDR()		read_sysreg(mpidr_el1)
+#define __get_MPIDR()		read_sysreg(mpidr_el1)
```
@andyross   >> Any reason for the oddball capitalization?
CMSIS (M/R/ v7A)  provides these definitions.  CMSIS for armv8-a is not available.
We will try to keep the needed definitions signatures same as CMSIS so that where they are used
they can be Cortex-A/M/R/V8A  agnostic.

stephanosio replied:
Unless we plan to switch the entire AArch64 CPU interface to CMSIS-style, IMHO, this should stay `GET_MPIDR`, and its users should do `#ifdef`.

This is one of the reasons why I suggested defining AArch64 symbols in CMSIS-style earlier.

sandeepbrcm replied:
Ideal is to have CMSIS like defines but Cortex-A Aarch64 CMSIS is not planned as per what I learned from arm tools team.
If more products use Cortex-A aarch64 as a high-end micro controller things can change (unlikely).

It also depends on where we need the abstraction.
ex: __DSB()  we have it defined as CMSIS. but really this should be abstracted at OS level then down the layer if its from CMSIS or some arch define would not matter.  Treating Aarch64 as a non-CMSIS port made sense so far.
But I am not against making all the defines CMSIS style, just that lack of time at the moment.

I can fallback to own define and add #ifdef at interrupt test routine. Because later point the gic_send_sgi()  itself needs an abstraction with cpu logical map if ever SMP support is added.

sandeepbrcm replied:
Use non-CMSIS for ARM64.

stephanosio reviewed (changes requested):
Thanks for looking into this.

This seems to a good starting point, but I wish we could end up with something more elegant -- I have left some comments below.

sandeepbrcm commented:
> Thanks for looking into this.
> 
> This seems to a good starting point, but I wish we could end up with something more elegant -- I have left some comments below.

Thanks. I am sure it will improve.

sandeepbrcm commented on drivers/interrupt_controller/intc_gic_common.c:
```diff
@@ -0,0 +1,153 @@
+
+	/* Disable the interrupt */
+	sys_write32(mask, ICENABLER(base, idx));
+#if defined(CONFIG_GIC_V3)
+	gic_wait_rwp(intid);
+#endif
```
ACK .  will update all places.

sandeepbrcm commented on include/drivers/interrupt_controller/gic.h:
```diff
@@ -51,166 +51,17 @@
  */
 #define	GICD_IIDR		(GIC_DIST_BASE +   0x8)
 
-/*
- * 0x080  Interrupt Group Registers
- * v1		ICDISRn
- * v2/v3	GICD_IGROUPRn
- */
-#define	GICD_IGROUPRn		(GIC_DIST_BASE +  0x80)
-
-/*
- * 0x100  Interrupt Set-Enable Reigsters
- * v1		ICDISERn
- * v2/v3	GICD_ISENABLERn
- */
-#define	GICD_ISENABLERn		(GIC_DIST_BASE + 0x100)
```
>to have some definitions here and others in intc_gic_common_priv.h

I actually intended to move all registers to private common header but forgot. I will move these remaining registers defines to` intc_gic_common_priv.h` .   There may be more things in the public header which do not belong here we can
clean them aswell.

>#define	GICD_IGROUPRn		(GIC_DIST_BASE +  0x80)

Defines in this form were not re-usable (recall the discussion https://github.com/zephyrproject-rtos/zephyr/pull/24850#discussion_r419206197 ) as they use absolute address hence the change. 
https://github.com/zephyrproject-rtos/zephyr/blob/38fba0c435f027a98692354549b037c035c50ae4/drivers/interrupt_controller/intc_gic_common_priv.h#L10-L34

sandeepbrcm replied:
Reorganized them further to move out of public header.

Commit:
arch: arm: api to get cpuid

Add api to get cpuid from MPIDR.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

Commit:
tests: kernel: interrupt: use GIC sgi api

Cleanup GIC sgi api usage. Use 'GET_CPU' and 'gic_raise_sgi'
to make the test more generic.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

sandeepbrcm commented on drivers/interrupt_controller/intc_gic_common.c:
```diff
@@ -0,0 +1,153 @@
+	#if CONFIG_GIC_VER >= 3
+	/* wait for rwp on GICD */
+	gic_wait_rwp(GIC_SPI_INT_BASE);
+	#endif
```
Done

MrVan commented:
Just has a question, has zephyr arm64 support smp mode?

sandeepbrcm commented:
> Just has a question, has zephyr arm64 support smp mode?

No.

MrVan commented:
> > Just has a question, has zephyr arm64 support smp mode?
> 
> No.

ok, so any plan to add smp support?

sandeepbrcm commented:
> > > Just has a question, has zephyr arm64 support smp mode?
> > 
> > 
> > No.
> 
> ok, so any plan to add smp support?

Very unlikely as per our Broadcom's requirements even in future.  If someone starts I can support reviews though.
